### ISSUE TITLE
Write Design and Architecture document

### ISSUE THREAD

**Issue by snehitp** [ClinGen/clinbeacon#1]
Please login with your gmail ID, or leave authored comments.
See document here: https://drive.google.com/open?id=0B_eulwICg5fRUW5xcUdBZTdGU0E

**Comment by snehitp** [comment]
Document contents need to be transferred to repo wiki once design choices are finalized.

**Comment by snehitp** [comment]
@trentmswanson I agree that it makes sense to flesh out the technical architecture parts that are domain-agnostic (e.g. tenancy, client responsibilities, centralized responsibilities, etc.). That way, your team can start working on a skeleton implementation that walks through the entire workflow with toy datasets before we start incorporating genomics-specific functionality.

**Comment by slowkoni** [comment]
@snehitp I'm confused by the use of the term domain-agnostic, can you clarify what distinction you are trying to make? 

I think also that if we use the term "federated", that doesn't mean separate database names in a single mongo/mysql/whatever, nor multiple mongos/mysqls/whatevers all running on the same system but listening on different ports, but rather what at least appears to be physically distinct systems that can only communicate via a layer 7 protocol we develop that is channeled over http (or something else if convenient). Each database in the federation would then have a distinct IP address with our client software running and listening on a specific TCP port (of our choosing, but then always the same so our server knows where to find it). Our client software would handle talking to the database server which is on "localhost" from it's perspective (and only accessible via localhost TCP connections or unix socket - not listening for outside network traffic). When we want to shift costs for hosting very large databases to member institutions that should have deep enough pockets to pay for it (ie, stanford), then we just bill the instance to them rather than us. This could all start as docker containers running on the same virtual instance, and then we move the containers to different instances as per any future considerations of the actual realized system load, space, or billing arrangements.

The access restrictions between member states of the federation is really simple though - it is defined at the IP level (firewall) to block anything going on other than our stuff, and our client software just rejects any query coming in via the layer 7 protocol that it doesn't understand, or doesn't allow. This should be very easy to explain to the IS/IT folks at any lab or institution. Otherwise, I think there may be considerable doubt from some places as to whether they are really separate from others and only these specified queries are executable.

I think trying to do anything not easily distinguished as separate hosts (and therefore separate IPs) is not really federated. If we don't do it this way or something similar, we'd only be telling them their data is separate, but really it isn't. If someone hacks our one instance, they'd have everyone's data all at once. It doesn't matter if this is done as docker containers or separate VM instances - that is a cost concern. Either docker or separate VMs should be easy to scale up or down. There may be another way than just docker or full instances.

Trent raises a good point about the actual physical data storage. Everybody might need separate EBS vols for database's underlying filesystem storage. In my experience, the major cost with cloud is usually the allocated EBS space. The meter runs on that all the time.

Oh, and we'll need service discovery - something like serf.

**Comment by snehitp** [comment]
@slowkoni These are all excellent technical points. I've revised the description of this issue just to be clear.

Could I ask you to structure your comments in a concise manner against @trentmswanson 's Google doc? It is unclear whether each of these corresponds to a feature request, and design decision or a general brainstorming idea. 

To make things actionable, please assign each point you raised as a comment - either to a subsection (single atomic features) or a section (multiple features grouped into a functionality category) on Trent's document. If a section or subsection does not already exist, then please create a new one. That way we can assign them a priority and a milestone at the atomic or the group level.

**Comment by trentmswanson** [comment]
We started working on the portal and and the beacon. I have a couple more assumptions I would like to validate regarding the federated approach, as well as questions. How much autonomy and control does the institution require for it's data?

__Centralized Identify and Access Management__
Although the data is decentralized and behind an API, Identity and access management will be centralized, **correct?** The API will only exposes the data that is needed by the system; run queries, import data methods, and basic (restricted view, delete, edit, update) methods for the data. The Clingen application will be responsible for controlling access to features like import, etc... based on the role, and authenticating the users. We could use authorization tokens to allow the client to talk directly to the beacon API, including the upload but the important point here is that identities and access will be centralized.

__Centralized import__
We could change this in the future, but for now I assume it's fine if the data is imported through the user portal (provided the user is an admin or submitter), processed and then imported into the tenant through the beacon API. We might be able to improve scale by allowing the data to flow from the client directly to the beacon, but we would still control access to the external API by generating a token for the tenant endpoint. This may be a v2 feature enhancement, and we could refactor things around this.

__Centralized data management services (same as above)__
A user will login to the same portal used to query data, and provided they are a submitter or admin they will have access to view, import, delete, etc... Once again is it Ok that this is a pass-through operation and that the beacon API exposes that to the centralized application or does the institution require a UI and complete control over the beacon data directly.  Maybe these are some v2 enhancements?  Basically we would be configuring the beacon using a trusted subsystem approach with the portal application and import services.

![image](https://cloud.githubusercontent.com/assets/1594373/16016745/56c3676e-3151-11e6-8969-36663a97d0f8.png)

**Comment by trentmswanson** [comment]
@slowkoni @snehitp will an individual patient have multiple samples in the VCF or does a sample represent one patient?  I just wanted to clarify this because I noticed the recent VCF sample document contained two samples in it.

**Comment by slowkoni** [comment]
It actually may be a common situation for us that a single person has multiple genomes - the most obvious example is tumor/normal sequencing where a cancer patient has the full genome or exome sequence done from DNA extracted from one or more biopsies of the tumor, and then their "normal" genome (called "germline genome" usually) sequenced as well for comparison. This sort of activity is actually where the bulk of full genome or full exome sequencing may soon be going on in clinical practice and where a shit-ton of our data might eventually come from. The reason for that is cancer is always a genetic disease and this is one area where it may be financially advantageous for insurance to pay for it. I do not know if we will ever see the germline genome and the tumor genome presented together as two different samples in the same VCF, though that would be awfully convenient for comparing them and that is the whole point of doing it - find out what changed. So far I have only seen it as separate, single-sample VCFs.

@snehitp Snehit what do you think about this? If our db can ever know that two de-identified genomes are actually the tumor-normal pair for the same patient, and of course the tumor type, this would make the database extremely powerful and informative for where all the money is going to be at in the next couple years.
